lms/djangoapps/instructor/views/tools.py

```diff
@@ -158,50 +150,33 @@ def title_or_url(node):
     return title
 
 
-def set_due_date_extension(course, unit, student, due_date):
+def set_due_date_extension(course, unit, student, due_date, actor=None):
     """
     Sets a due date extension. Raises DashboardError if the unit or extended
     due date is invalid.
     """
     if due_date:
-        # Check that the new due date is valid:
-        with disable_overrides():
-            original_due_date = getattr(unit, 'due', None)
-
-        if not original_due_date:
+        try:
+            api.set_date_for_block(course.id, unit.location, 'due', due_date, user=student, reason=None, actor=actor)
+        except api.MissingDateError:
             raise DashboardError(_(u"Unit {0} has no due date to extend.").format(unit.location))
-        if due_date < original_due_date:
+        except api.InvalidDateError:
             raise DashboardError(_("An extended due date must be later than the original due date."))
-
-        override_field_for_user(student, unit, 'due', due_date)
-
     else:
-        # We are deleting a due date extension. Check that it exists:
-        if not get_override_for_user(student, unit, 'due'):
-            raise DashboardError(_("No due date extension is set for that student and unit."))
-
-        clear_override_for_user(student, unit, 'due')
+        api.set_date_for_block(course.id, unit.location, 'due', None, user=student, reason=None, actor=actor)
 
 
 def dump_module_extensions(course, unit):
     """
     Dumps data about students with due date extensions for a particular module,
     specified by 'url', in a particular course.
     """
-    data = []
     header = [_("Username"), _("Full Name"), _("Extended Due Date")]
-    query = StudentFieldOverride.objects.filter(
-        course_id=course.id,
-        location=unit.location,
-        field='due')
-    for override in query:
-        due = DATE_FIELD.from_json(json.loads(override.value))
-        due = due.strftime(u"%Y-%m-%d %H:%M")
-        fullname = override.student.profile.name
-        data.append(dict(zip(
-            header,
-            (override.student.username, fullname, due))))
-    data.sort(key=lambda x: x[header[0]])
+    data = []
+    for username, fullname, due_date in api.get_overrides_for_block(course.id, unit.location):
+        due_date = due_date.strftime(u'%Y-%m-%d %H:%M')
+        data.append(dict(zip(header, (username, fullname, due_date))))
+    data.sort(key=operator.itemgetter(_("Username")))
```

Comment:
Okay, I have lots of questions about this, but I see that you just moved it around and cleaned it up a bit as opposed to writing it originally, so that'll be for another time. :-/

Reply:
Yes, there's room for improvement here...